tests/gradients/test_parameter_shift.py

```diff
@@ -287,22 +287,31 @@ def test_with_gradient_recipes(self):
         assert tapes[3].get_parameters(trainable_only=False) == [1.0, 2.0, 2 * 3.0 + 2, 4.0]
         assert tapes[4].get_parameters(trainable_only=False) == [1.0, 2.0, 3 * 3.0 + 3, 4.0]
 
-    def test_recycled_unshifted_tape(self):
+    @pytest.mark.parametrize("ops_with_custom_recipe", [[0], [1], [0, 1]])
+    def test_recycled_unshifted_tape(self, ops_with_custom_recipe):
         """Test that if the gradient recipe has a zero-shift component, then
         the tape is executed only once using the current parameter
         values."""
         dev = qml.device("default.qubit", wires=2)
 
         with qml.tape.QuantumTape() as tape:
             qml.RX(0.543, wires=[0])
-            qml.RY(-0.654, wires=[0])
+            qml.RX(-0.654, wires=[0])
             qml.expval(qml.PauliZ(0))
 
-        gradient_recipes = ([[-1e7, 1, 0], [1e7, 1, 1e7]],) * 2
+        gradient_recipes = tuple(
+            [[-1e7, 1, 0], [1e7, 1, 1e-7]] if i in ops_with_custom_recipe else None
+            for i in range(2)
+        )
         tapes, fn = qml.gradients.param_shift(tape, gradient_recipes=gradient_recipes)
 
-        # one tape per parameter, plus one global call
-        assert len(tapes) == tape.num_params + 1
+        # two tapes per parameter that doesn't use a custom recipe,
+        # one tape per parameter that uses custom recipe,
+        # plus one global call if at least one uses the custom recipe

```

Comment:
Based on this comment, not sure: how come we have `- len(ops_with_custom_recipe)` in line 311?

Reply:
Ah, it is 
`2*(tape.num_params - len(ops_with_custom_recipe))` <- operations without custom recipe each get two tapes
`len(ops_with_custom_recipe)` <- operations with custom recipe each get one tape
`+1` for the unshifted tape.
Adding those, we get 
`2*(tape.num_params - len(ops_with_custom_recipe))+len(ops_with_custom_recipe)+1`
`=2*tape.num_params - len(ops_with_custom_recipe)+1`

Reply:
Oh okay, that makes sense, thank you :smile:

Could we then have
```py
num_op_no_custom_recipes = tape.num_params - len(ops_with_custom_recipe)
assert len(tapes) == 2 * num_op_no_custom_recipes + len(ops_with_custom_recipe) + 1
```
?

It sacrifices some arithmetic simplification in return for reflecting the comment term by term.

Reply:
Great call! Done :)